torch/distributions/kumaraswamy.py

```diff
@@ -0,0 +1,67 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.uniform import Uniform
+from torch.distributions.transformed_distribution import TransformedDistribution
+from torch.distributions.transforms import AffineTransform, PowerTransform
+from torch.distributions.utils import broadcast_all, euler_constant, _finfo
+
+
+def _moments(a, b, n):
+    """
+    Computes nth moment of Kumaraswamy using using torch.lgamma
+    """
+    arg1 = 1 + n / a
+    log_value = torch.lgamma(arg1) + torch.lgamma(b) - torch.lgamma(arg1 + b)
+    return b * torch.exp(log_value)
+
+
+class Kumaraswamy(TransformedDistribution):
+    r"""
+    Samples from a Kumaraswamy distribution.
+
+    Example::
+
+        >>> m = Kumaraswamy(torch.Tensor([1.0]), torch.Tensor([1.0]))
+        >>> m.sample()  # sample from a Kumaraswamy distribution with concentration alpha=1 and beta=1
+        tensor([ 0.1729])
+
+    Args:
+        concentration1 (float or Tensor): 1st concentration parameter of the distribution
+            (often referred to as alpha)
+        concentration0 (float or Tensor): 2nd concentration parameter of the distribution
+            (often referred to as beta)
+    """
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
+    support = constraints.unit_interval
+    has_rsample = True
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        self.concentration1, self.concentration0 = broadcast_all(concentration1, concentration0)
+        finfo = _finfo(self.concentration0)
+        base_dist = Uniform(torch.full_like(self.concentration0, finfo.tiny),
+                            torch.full_like(self.concentration0, 1 - finfo.eps))
+        transforms = [AffineTransform(loc=1., scale=-torch.ones_like(self.concentration1)),
+                      PowerTransform(exponent=self.concentration0.reciprocal()),
+                      AffineTransform(loc=1., scale=-torch.ones_like(self.concentration0)),
```

Comment:
It's so bizarre why this is happening: the base distribution samples points in the open interval (0, 1), and the first two transforms work correctly by maintaining this range. However, this `AffineTransform` seems to make certain values close to 0 = 1, which is causing the entropy test failure.

Any thoughts on how to mitigate this issue?

Reply:
Change the `.tiny` to `.eps`. This is because `eps` is around 1e-8 but tiny is around 1e-38. Floating point representation can measure very tiny differences near 0, but only 1e-8 differences near 1.

Reply:
I tried this earlier, and it didn't seem to work. Even then, it was the same transform causing the problem too.

Reply:
Hmm you could try changing it to `self.concentration * eps`, to ensure it's still measurable after the `PowerTransform`?

Reply:
```diff
-        base_dist = Uniform(torch.full_like(self.concentration0, finfo.tiny),
+        base_dist = Uniform(torch.full_like(self.concentration0, finfo.eps),
                             torch.full_like(self.concentration0, 1 - finfo.eps))
         transforms = [AffineTransform(loc=1., scale=-torch.ones_like(self.concentration1)),
                       PowerTransform(exponent=self.concentration0.reciprocal()),
-                      AffineTransform(loc=1., scale=-torch.ones_like(self.concentration0)),
+                      AffineTransform(loc=1., scale=-torch.full_like(self.concentration0, 1 - finfo.eps)),
```

Sorry @fritzo , even this doesn't seem to help? Is the diff what you had in mind?

Reply:
`eps` is about 1e-08, and if we were to assume `self.concentration0` = 1e08 (hypothetically, of course, but still a valid parameter value), then the base distribution will get affected, won't it?

Reply:
However, even that doesn't seem to help.

Reply:
1. I think we can safely ignore extreme concentration like `concentration0=1-e8`, or to require users to use double precision in extreme cases.
2. For non-extreme cases, it shouldn't matter that the base distribution is affected, since `PowTransform` will map the smallest value back down to `eps`. Of course there will be mass missing at 0, but we've already decided to ignore mass at discrete points.

Reply:
Thanks @fritzo , but your solution doesn't seem to work. This is how I am finding out the bug:
```diff
                 expected = -dist.log_prob(x).mean(0)
                 ignore = (expected == inf)
                 expected[ignore] = actual[ignore]
+                if torch.any(torch.isnan(expected)):
+                    x = dist.base_dist.sample(sample_shape=(60000,))
+                    for trans in dist.transforms:
+                        x = trans(x)
+                        print(trans)
+                        print(torch.any(x >= 1))
+                        print(torch.any(x <= 0))
```

This is the captured output by `pytest`:
```
AffineTransform()
tensor(0, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
PowerTransform()
tensor(0, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
AffineTransform()
tensor(1, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
PowerTransform()
tensor(1, dtype=torch.uint8)
tensor(0, dtype=torch.uint8)
```

Reply:
Ok, why don't you compute the first-order scaling of `eps` that is needed to make it through all transforms. It should involve `concentration0` and `concentration1`.